core/kernel/interrupt.c

```diff
@@ -37,51 +40,196 @@
 	}
 }
 
-void itr_add(struct itr_handler *h)
+static itr_result add_handler(size_t it, itr_handler handler,
+		uint32_t flags, void *data)
+{
+	itr_result res;
+	struct itr_desc *desc = itr_descs[it];
+	struct itr_desc *in_desc;
+	struct itr_desc *tmp_desc;
+	uint32_t old_itr_status;
+
+	in_desc = malloc(sizeof(struct itr_desc));
+	if (in_desc == NULL) {
+		EMSG("Fail to malloc.");
+		return GIC_BAD_MEMORY;
+	}
+	in_desc->it = it;
+	in_desc->flags = flags;
+	in_desc->handler = handler;
+	in_desc->data = data;
+	in_desc->d_next = NULL;
+
+	old_itr_status = thread_mask_exceptions(THREAD_EXCP_ALL);
+	cpu_spin_lock(&itr_lock);
+	if (desc == NULL) {
+		/* shared interrupt's data should be not null */
+		if ((flags & ITRF_SHARE_MASK) != ITRF_SHARE || data != NULL) {
+			itr_descs[it] = in_desc;
+			res = GIC_OK;
+		} else {
+			EMSG("The data should not null when itr is shared.");
+			res = GIC_BAD_PARAMETER;
+		}
+		goto exit;
+	}
+	if ((desc->flags & ITRF_SHARE_MASK) == (flags & ITRF_SHARE_MASK)
+		&& ((desc->flags & ITRF_SHARE_MASK) == ITRF_SHARE)) {
+		/* the data should be unique */
+		if (data == NULL || (size_t)desc->data == (size_t)data) {
+			EMSG("The handler data is not valid.");
+			res = GIC_BAD_PARAMETER;
+			goto exit;
+		}
+
+		tmp_desc = desc;
+		desc = desc->d_next;
+		for (; desc; desc = desc->d_next) {
+			if ((size_t)desc->data == (size_t)data) {
+				EMSG("The handler data is not unique.");
+				res = GIC_BAD_PARAMETER;
+				goto exit;
+			}
+			tmp_desc = desc;
+		}
+		tmp_desc->d_next = in_desc;
+		res = GIC_OK;
+	}
+
+exit:
+	if (res != GIC_OK)
+		free(in_desc);
+	cpu_spin_unlock(&itr_lock);
+	thread_unmask_exceptions(old_itr_status);
+	return res;
+}
+
+static itr_result remove_handler(size_t it, uint32_t flags, void *data)
+{
+	itr_result res;
+	struct itr_desc *desc = itr_descs[it];
+	struct itr_desc *tmp_desc;
+	uint32_t old_itr_status;
+
+	old_itr_status = thread_mask_exceptions(THREAD_EXCP_ALL);
+	cpu_spin_lock(&itr_lock);
+	if (desc == NULL) {
+		res = GIC_OK;
+		goto exit;
+	}
+
+	if ((desc->flags & ITRF_SHARE_MASK) == (flags & ITRF_SHARE_MASK)
+		&& (desc->flags & ITRF_SHARE_MASK) == ITRF_SHARE) {
+		if (data == NULL) {
+			EMSG("The hanlder data should be not null.");
+			res = GIC_BAD_PARAMETER;
+			goto exit;
+		}
+		if ((size_t)desc->data == (size_t)data) {
+			itr_descs[it] = desc->d_next;
+			res = GIC_OK;
+			goto exit;
+		}
+
+		tmp_desc = desc;
+		desc = desc->d_next;
+		for (; desc; desc = desc->d_next) {
+			if ((size_t)desc->data == (size_t)data) {
+				tmp_desc->d_next = desc->d_next;
+				res = GIC_OK;
+				goto exit;
+			}
+			tmp_desc = desc;
+		}
+	} else
+		itr_descs[it] = NULL;
+
+	res = GIC_OK;
+exit:
+	if (desc != NULL)
+		free(desc);
+	cpu_spin_unlock(&itr_lock);
+	thread_unmask_exceptions(old_itr_status);
+	return res;
+}
+
+itr_result itr_add(size_t it, itr_handler handler, uint32_t flags, void *data)
+{
+	itr_chip->ops->add(itr_chip, it, flags);
+	return add_handler(it, handler, flags, data);
```

Comment:
i would propose the other way, to handle error case:
```
        itr_result ret;

        ret = add_handler(it, handler, flags, data);
        if (ret == GIC_OK)
                itr_chip->ops->add(itr_chip, it, flags);
        return ret;

```
if add_handler() failed, the it should be not be